File: dissertation/sim/two_particle/input.py
from pathlib import Path
from uuid import UUID

from pydantic import BaseModel, Field

THIS_DIR = Path(__file__).parent


class ParticleInput(BaseModel):
    id: UUID
    x: float = 0
    y: float = 0
    radius: float = Field(ge=0)
    ovality: float = Field(ge=0, lt=1, default=0)
    peak_count: int = Field(ge=0, default=0)
    peak_height: float = Field(ge=0, lt=1, default=0)
    node_count: int = Field(gt=0, default=100)


class InterfaceInput(BaseModel):
    energy: float = Field(gt=0)
    diffusion_coefficient: float = Field(gt=0)


class MaterialInput(BaseModel):
    surface: InterfaceInput
    density: float = Field(gt=0)
    molar_mass: float = Field(gt=0)


class Input(BaseModel):
    particle1: ParticleInput
    particle2: ParticleInput
    material1: MaterialInput
    material2: MaterialInput
    grain_boundary: InterfaceInput
    gas_constant: float
    temperature: float
    duration: float
    vacancy_concentration: float

    @property
    def _time_norm_common(self):
        return (
            self.gas_constant
            * self.temperature
            / self.vacancy_concentration
            * self.particle1.radius**4
            / self.material1.molar_mass
            * self.material1.density
        )

    @property
    def time_norm_surface(self):
        return self._time_norm_common / (self.material1.surface.diffusion_coefficient * self.material1.surface.energy)

    @property
    def time_norm_grain_boundary(self):
        return self._time_norm_common / (self.grain_boundary.diffusion_coefficient * self.grain_boundary.energy)

File: dissertation/sim/two_particle/Input.cs
namespace parameter_study;

class ParticleInput
{
    public required Guid Id { get; set; }
    public required double X { get; set; }
    public required double Y { get; set; }
    public required double RotationAngle { get; set; }
    public required double Radius { get; set; }
    public required double Ovality { get; set; }
    public required int PeakCount { get; set; }
    public required double PeakHeight { get; set; }
    public required int NodeCount { get; set; }
}

class InterfaceInput
{
    public required double Energy { get; set; }
    public required double DiffusionCoefficient { get; set; }
}

class MaterialInput
{
    public required double Density { get; set; }
    public required double MolarMass { get; set; }
    public required InterfaceInput Surface { get; set; }
}

class Input
{
    public required ParticleInput Particle1 { get; set; }
    public required ParticleInput Particle2 { get; set; }
    public required MaterialInput Material1 { get; set; }
    public required MaterialInput Material2 { get; set; }
    public required InterfaceInput GrainBoundary { get; set; }

    public required double Temperature { get; set; }

    public required double VacancyConcentration { get; set; }

    public required double GasConstant { get; set; }

    public required double Duration { get; set; }

    public required FreeSurfaceRemesherOptions? FreeSurfaceRemesherOptions { get; set; }

    public required double NeckDeletionLimit { get; set; }

    public required double TimeStepAngleLimit { get; set; }
}

class FreeSurfaceRemesherOptions
{
    public required double DeletionLimit { get; set; }

    public required double AdditionLimit { get; set; }

    public required double MinWidthFactor { get; set; }

    public required double MaxWidthFactor { get; set; }

    public required double TwinPointLimit { get; set; }
}
